### PR TITLE
YJIT: polymorphic opt_send_without_block

### PR DESCRIPTION
**Pull request by XrXr**
This gets rid of our dependency on the interpreter's call cache, which allows us to run well with `--yjit-callthreshold=1`. Polymorphism is just icing on the cake. Some refactoring to make doing guard chains is also included. They will be useful for other codgen functions.

Since this PR is quite large, I suggest going commit by commit to make review more manageable.

### CODE REVIEW

**Comment by XrXr** [comment]
> test_ractor.rb             ...................................Fstderr output is not empty
bootstraptest.tmp.rb:11: [BUG] rb_vm_insn_addr2opcode: invalid insn address: 0x000055a57155bdb1
https://github.com/Shopify/ruby/pull/83/checks?check_run_id=2177027195#step:12:31
   
Oh hmm, I guess this is means I should get serious about #80. Also I think patching iseqs when taking side exits without taking the VM lock is racy, so maybe I should get serious about [taking exits without patching iseqs](https://github.com/Shopify/ruby/commit/9fa855fdad1ff7f2b205b7e73e711704a9911b33).

**Comment by maximecb** [comment]
> Oh hmm, I guess this is means I should get serious about #80. Also I think patching iseqs when taking side exits without taking the VM lock is racy, so maybe I should get serious about taking exits without patching iseqs.

Lol at racy 😆 

Those both seem like important implementation obstacles to solve 👍

File: vm_method.c [hunk 1]
```diff
@@ -295,6 +297,8 @@ void
 rb_clear_method_cache_all(void)
 {
     rb_objspace_each_objects(invalidate_all_cc, NULL);
+
+    rb_yjit_invalidate_all_method_lookup_assumptions();
```
Comment (maximecb) [inline vm_method.c:301]: Oof, when and how often is this typically called during the execution of a Ruby program?

File: yjit_codegen.c [hunk 2]
```diff
@@ -1323,10 +1314,45 @@ gen_jump(jitstate_t* jit, ctx_t* ctx)
...
+
+    // Check that the receiver is a heap object
+    {
+        test(cb, REG0, imm_opnd(RUBY_IMMEDIATE_MASK));
+        jnz_ptr(cb, side_exit);
+        cmp(cb, REG0, imm_opnd(Qfalse));
+        je_ptr(cb, side_exit);
+        cmp(cb, REG0, imm_opnd(Qnil));
+        je_ptr(cb, side_exit);
+    }
```
Comment (maximecb) [inline yjit_codegen.c:1346]: Why do you insert blocks between curly braces like this one? There isn't a need for a new local scope here?

File: yjit_codegen.c [hunk 3]
```diff
@@ -1323,10 +1314,45 @@ gen_jump(jitstate_t* jit, ctx_t* ctx)
...
+    x86opnd_t klass_opnd = mem_opnd(64, REG0, offsetof(struct RBasic, klass));
+
+    // Bail if receiver class is different from compile-time call cache class
+    jit_mov_gc_ptr(jit, cb, REG1, known_klass);
+    cmp(cb, klass_opnd, REG1);
+    jit_chain_guard(JCC_JNE, jit, recompile_context, max_chain_depth, side_exit);
+    return true;
+}
+
+// Generate ancestry guard for protected callee.
```
Comment (maximecb) [inline yjit_codegen.c:1358]: Need a better explanation of what the generated code is checking for. I have no idea what a protected callee ancestry guard is, it's not an intuitive concept if you're not deeply familiar with CRuby's implementation.

**Comment by maximecb** [comment]
This is a huge PR 😲 but it seems like an improvement in terms of making the JIT code cleaner. I added some comments, most of them minor.

My main concern is how often `rb_clear_method_cache_all(void)` and ` rb_yjit_invalidate_all_method_lookup_assumptions();` gets called in practice. Because that's going to invalidate all of our method call blocks... And that's basically most of the blocks we generate, or at least a huge proportion, which is going to have big overhead and reduce the quality of our compiled code.

**Comment by maximecb** [comment]
Maybe we should also take this PR as an opportunity to make `--yjit-callthreshold=1` the default, or at least to make sure that the threshold is set to 1 when running tests?

File: vm_method.c (hunk 1 above)
Comment (XrXr) [reply vm_method.c:301]: Right now this is a noop. It seems that the interpreter doesn't need to do this either. I sent a PR upstream to remove the code and all tests pass: https://github.com/ruby/ruby/pull/4323

File: yjit_codegen.c (hunk 2 above)
Comment (XrXr) [reply yjit_codegen.c:1346]: Just for aesthetics. Do you prefer removing it?

Comment (maximecb) [reply yjit_codegen.c:1346]: In the future yes, but I will merge this PR now :)

**Pull request merged**